File: ts_sdk/src/models/LabelInfo.ts
// Label delimiter constant
export const LABEL_DELIMITER = '.';

/**
 * Ontology names enum - Single source of truth for all ontology identifiers
 * Use these constants instead of magic strings to ensure consistency
 */
export enum OntologyNames {
  SKILL = 'skill',
  GOOGLE = 'google',
  TASK = 'task',
  GOAL = 'goal',
}

/**
 * LabelInfo represents a label with metadata.
 * Labels can be:
 * - Ontology labels: "--<ontology>--.<path>" (e.g., "--skill--.solution_engineer")
 * - Ad-hoc labels: plain strings without ontology prefix (e.g., "manual")
 */
export class LabelInfo {
  constructor(
    public label: string,
    public description?: string,
    public color?: string | null,
    public parent?: string,
  ) {}

  /**
   * Get display name (last segment of label path)
   * Examples:
   * - "--skill--.solution_engineer" => "solution_engineer"
   * - "--google--.drive.upload" => "upload"
   * - "manual" => "manual"
   */
  get display(): string {
    return this.label.split(LABEL_DELIMITER).pop() || this.label;
  }

  /**
   * Parse a label ID into ontology name and path.
   * Format: "--<ontology>--.<path>"
   *
   * Examples:
   * - "--skill--.solution_engineer" => { ontology: "skill", path: "solution_engineer" }
   * - "--google--.drive.upload" => { ontology: "google", path: "drive.upload" }
   * - "manual" => { ontology: null, path: "manual" }
   *
   * @param labelId Full label ID
   * @returns Object with ontology name (null for ad-hoc) and path
   */
  static parseLabel(labelId: string): { ontology: string | null; path: string } {
    const match = labelId.match(/^--([^-]+)--\.(.+)$/);
    if (!match) {
      // Ad-hoc label without ontology prefix
      return { ontology: null, path: labelId };
    }
    return { ontology: match[1], path: match[2] };
  }
}

export class LabelInfoHelper {
  /**
   * Return the last segment of the label as the 'keyword'.
   */
  static getKeyword(labelInfo: LabelInfo): string {
    return labelInfo.label.split(LABEL_DELIMITER).pop() || '';
  }
}

File: flow_sdk/knowledge_engine/ontology/label_info.py
"""
LabelInfo class for managing individual labels in an ontology.
"""

from typing import Optional

from pydantic import BaseModel

# Label delimiter constant
LABEL_DELIMITER = "."


class LabelInfo(BaseModel):
    """Represents a single label in an ontology with metadata."""

    label: str  # e.g., "google.drive.upload"
    description: Optional[str] = None
    parent: Optional[str] = None
    color: Optional[str] = None  # Hex color code for UI representation

    @property
    def keyword(self) -> str:
        """Return the last segment of the label as the 'keyword'."""
        return self.label.split(LABEL_DELIMITER)[-1]

    @property
    def segments(self) -> list[str]:
        """Return all segments of the label."""
        return self.label.split(LABEL_DELIMITER)

    @property
    def parent_label(self) -> Optional[str]:
        """Return the parent label (all segments except the last)."""
        segments = self.segments
        if len(segments) <= 1:
            return None
        return LABEL_DELIMITER.join(segments[:-1])

    def is_child_of(self, parent_label: str) -> bool:
        """Check if this label is a child of the given parent label."""
        return self.label.startswith(f"{parent_label}{LABEL_DELIMITER}")

    def get_depth(self) -> int:
        """Return the depth of this label (number of segments)."""
        return len(self.segments)
